kernel/thread.c

```diff
@@ -348,40 +410,172 @@ long myst_syscall_wait4(
+                    printf(
+                        "    WIFSTOPPED (signaled to stop, not supported): "
+                        "%s\n",
+                        WIFSTOPPED(*wstatus) ? "true" : "false");
+                    if (WIFSTOPPED(*wstatus))
+                        printf(
+                            "    WSTOPSIG (signal code to stop, not "
+                            "supported): "
+                            "%u\n",
+                            WSTOPSIG(*wstatus));
+                    printf(
+                        "    WIFCONTINUED (signaled to resumed after stop, not "
+                        "supported): %s\n",
+                        WIFCONTINUED(*wstatus) ? "true" : "false");
+#endif
+                }
                 ret = p->pid;
+
+                // remove from zombie list
+                if (p->zprev == NULL)
+                {
+                    _zombies_head = p->znext;
+                }
+                else
+                {
+                    p->zprev->znext = p->znext;
+                }
+                if (p->znext != NULL)
+                    p->znext->zprev = p->zprev;
+
+                if (_zombies_tail == p)
+                    _zombies_tail = p->group_prev;
+
+                _zombies_count--;
+
+                // remove from process list
+                myst_spin_lock(&myst_process_list_lock);
+                if (p->main.prev_process_thread)
+                    p->main.prev_process_thread->main.next_process_thread =
+                        p->main.next_process_thread;
+                if (p->main.next_process_thread)
+                    p->main.next_process_thread->main.prev_process_thread =
+                        p->main.prev_process_thread;
+                myst_spin_unlock(&myst_process_list_lock);
+
+                // free zombie thread
+                free(p);
+
                 goto done;
             }
         }
 
+        /* no zombies found, but are there any running childred? If not then
+         * ECHILD returned */
+        myst_spin_lock(&myst_process_list_lock);
+
+        // first processes left
+        p = process->main.prev_process_thread;
+        while (p && !_wait_matcher("active", pid, process, p))
+        {
+            p = p->main.prev_process_thread;
+        }
+        if (p == NULL)
+        {
+            // now processes right
+            p = process->main.next_process_thread;
+            while (p && !_wait_matcher("active", pid, process, p))
+            {
+                p = p->main.next_process_thread;
+            }
+        }
+        if (p == NULL)
+        {
+#ifdef TRACE
+            printf("Returning ECHILD\n");
+#endif
+            myst_spin_unlock(&myst_process_list_lock);
+            ERAISE(-ECHILD);
+        }
+
+        myst_spin_unlock(&myst_process_list_lock);
+
         if ((options & WNOHANG))
         {
             ret = 0;
             goto done;
         }
 
         /* wait for signal from myst_zombify_thread() */
```

Comment:
The comment should be removed also. We are not using conditional variables now.

Reply:
done